net/portmapper/probe.go

```diff
@@ -0,0 +1,332 @@
+		defer p.PMP.Set(false, nil)
+		defer p.PCP.Set(false, nil)
+		for {
+			pmp_ctx, cancel := context.WithTimeout(ctx, portMapServiceTimeout)
+			hasPCP, hasPMP, err := c.probePMPAndPCP(pmp_ctx)
+			if err != nil {
+				if ctx.Err() != nil {
+					err = nil
+					// the global context has passed, exit cleanly
+					cancel()
+					return
+				}
+				if pmp_ctx.Err() == context.DeadlineExceeded {
+					err = nil
+				}
+			}
+			cancel()
+			p.PMP.Set(hasPMP, err)
+			p.PCP.Set(hasPCP, err)
+
+			t := time.NewTimer(trustServiceStillAvailableDuration * 3 / 4)
+
+			select {
+			case should_pause := <-pause:
+				if !should_pause {
+					t.Stop()
+					return
+				}
+				restart := <-pause
+				if !restart {
+					t.Stop()
+					return
+				}
+			case <-t.C: // break through and retry the connection
+			}
+		}
+	}()
+
+	go func() {
+		defer p.UPnP.Set(false, nil)
+		for {
+			upnp_ctx, cancel := context.WithTimeout(ctx, portMapServiceTimeout)
+			hasUPnP, err := c.probeUPnP(upnp_ctx)
+			if err != nil {
+				if ctx.Err() != nil {
+					// the global context has passed, exit cleanly
+					cancel()
+					return
+				}
+				if upnp_ctx.Err() == context.DeadlineExceeded {
+					err = nil
+				}
+			}
+			cancel()
+			t := time.NewTimer(trustServiceStillAvailableDuration * 3 / 4)
+			p.UPnP.Set(hasUPnP, err)
+
+			select {
+			case should_pause := <-pause:
+				if !should_pause {
+					t.Stop()
+					return
+				}
+				restart := <-pause
+				if !restart {
+					t.Stop()
+					return
+				}
+			case <-t.C: // break through and retry the connection
+			}
+		}
+	}()
+
+	return
+}
+
+// Stop gracefully turns the Prober off, completing the current probes before exiting.
+func (p *Prober) Stop() { close(p.pause) }
+
+// Pauses the prober if currently running, or starts if it was previously paused.
+func (p *Prober) Toggle() { p.pause <- true }
+
+// CurrentStatus returns the current results of the prober, regardless of whether they have
+// completed or not.
+func (p *Prober) CurrentStatus() (res ProbeResult, err error) {
+	hasPMP, errPMP := p.PMP.PresentCurrent()
+	res.PMP = hasPMP
+	err = errPMP
+
+	hasUPnP, errUPnP := p.UPnP.PresentCurrent()
+	res.UPnP = hasUPnP
+	if err == nil {
+		err = errUPnP
+	}
+
+	hasPCP, errPCP := p.PCP.PresentCurrent()
+	res.PCP = hasPCP
+	if err == nil {
+		err = errPCP
+	}
+	return
+}
+
+func (p *Prober) StatusBlock() (res ProbeResult, err error) {
+	hasPMP, errPMP := p.PMP.PresentBlock()
+	res.PMP = hasPMP
+	err = errPMP
+
+	hasUPnP, errUPnP := p.UPnP.PresentBlock()
+	res.UPnP = hasUPnP
+	if err == nil {
+		err = errUPnP
+	}
+
+	hasPCP, errPCP := p.PCP.PresentBlock()
+	res.PCP = hasPCP
+	if err == nil {
+		err = errPCP
+	}
+	return
+}
+
+type ProbeSubResult struct {
+	cond *sync.Cond
+	// If this probe has finished, regardless of success or failure
+	completed bool
+
+	// whether or not this feature is present
+	present bool
+	// most recent error
+	err error
+
+	// time we last saw it to be available.
+	sawTime time.Time
```

Comment:
what are all these guarded by?

Reply:
they're guarded by the lock inside of the `Cond`, what's the best way to handle that?